src/msg/async/frames_v2.cc

```diff
@@ -161,6 +172,87 @@ void FrameAssembler::asm_secure_r0(const preamble_block_t& preamble,
   append_encrypt_final(frame_bl);
 }
 
+void FrameAssembler::asm_crc_r1(const preamble_block_t& preamble,
+                                bufferlist* segment_bls,
+                                bufferlist* frame_bl) const {
+  frame_bl->append(reinterpret_cast<const char*>(&preamble),
+                   sizeof(preamble));
+
+  epilogue_crc_r1_block_t epilogue;
+  // FIPS zeroization audit 20191115: this memset is not security related.
+  ::memset(&epilogue, 0, sizeof(epilogue));
+  epilogue.late_status |= FRAME_LATE_STATUS_COMPLETE;
+
+  ceph_assert(segment_bls[0].length() == m_descs[0].logical_len);
+  if (segment_bls[0].length() > 0) {
+    auto crc = init_le32(segment_bls[0].crc32c(-1));
+    frame_bl->claim_append(segment_bls[0]);
```

Comment:
Ah, that's one the reasons behind the questions about `bufferlist`.

Reply:
Yup.